lib/scala/interpreter-dsl/src/main/java/org/enso/interpreter/dsl/BuiltinsProcessor.java

```diff
@@ -0,0 +1,211 @@
+public class BuiltinsProcessor extends AbstractProcessor {
+
+  private static final String BuiltinsPkg = "org.enso.interpreter.node.expression.builtin";
+
+  @Override
+  public final boolean process(Set<? extends TypeElement> annotations, RoundEnvironment roundEnv) {
+    for (TypeElement annotation : annotations) {
+      Set<? extends Element> annotatedElements = roundEnv.getElementsAnnotatedWith(annotation);
+      for (Element elt : annotatedElements) {
+        if (elt.getKind() == ElementKind.METHOD) {
+          try {
+            handleMethodElement(elt, roundEnv);
+          } catch (IOException ioe) {
+            ioe.printStackTrace();
+          }
+        } else {
+          throw new RuntimeException(
+              "Invalid use of " + annotation.getSimpleName() + " with " + elt.getKind());
+        }
+      }
+    }
+    return true;
+  }
+
+  public void handleMethodElement(Element element, RoundEnvironment roundEnv) throws IOException {
+    ExecutableElement method = (ExecutableElement) element;
+    String methodName = method.getSimpleName().toString();
+    Element owner = element.getEnclosingElement();
+    if (owner.getKind() == ElementKind.CLASS) {
+      TypeElement tpeElement = (TypeElement) owner;
+      PackageElement pkgElement = (PackageElement) tpeElement.getEnclosingElement();
+      Builtin annotation = element.getAnnotation(Builtin.class);
+
+      String builtinPkg =
+          annotation.pkg().isEmpty() ? BuiltinsPkg : BuiltinsPkg + "." + annotation.pkg();
+      ClassName builtinMethodNode =
+          new ClassName(
+              builtinPkg,
+              methodName.substring(0, 1).toUpperCase() + methodName.substring(1) + "Node");
+      ClassName ownerClass =
+          new ClassName(pkgElement.getQualifiedName(), tpeElement.getSimpleName());
+      JavaFileObject gen =
+          processingEnv.getFiler().createSourceFile(builtinMethodNode.fullyQualifiedName());
+      MethodGenerator methodGen = new MethodGenerator(method);
+      generateBuiltinMethodNode(
+          gen, methodGen, methodName, annotation.description(), builtinMethodNode, ownerClass);
+    } else {
+      throw new RuntimeException("@Builtin method must be owned by the class");
+    }
+  }
+
+  private void generateBuiltinMethodNode(
+      JavaFileObject gen,
+      MethodGenerator mgen,
+      String methodName,
+      String description,
+      ClassName builtinNode,
+      ClassName ownerClazz)
+      throws IOException {
+    String ensoMethodName = methodName.replaceAll("([^_A-Z])([A-Z])", "$1_$2").toLowerCase();
+    try (PrintWriter out = new PrintWriter(gen.openWriter())) {
+      out.println("package " + builtinNode.getPkgName() + ";");
+      out.println();
+      for (String importPkg : necessaryImports) {
+        out.println("import " + importPkg + ";");
+      }
+      out.println("import " + ownerClazz.fullyQualifiedName() + ";");
+      out.println();
+      out.println(
+          "@BuiltinMethod(type = \""
```

Comment:
A processor that generates code for another processor - that's _metacyclic_!

Reply:
![image](https://user-images.githubusercontent.com/292128/167824050-11e1f343-6746-4d17-9e36-cd70e7d63765.png)